core/transactions/staged_mutation.cxx

```diff
@@ -22,9 +22,46 @@
 #include "internal/utils.hxx"
 #include "result.hxx"
 
+#include <asio/bind_executor.hpp>
+#include <asio/post.hpp>
+
 namespace couchbase::core::transactions
 {
 
+bool
+unstaging_state::wait_until_unstage_possible()
+{
+    std::unique_lock lock(mutex_);
+    auto timeout = std::max(std::chrono::nanoseconds(0), ctx_->overall().remaining()) + timeout_defaults::key_value_durable_timeout +
+                   std::chrono::seconds(1);
+    auto success = cv_.wait_for(lock, timeout, [this] { return (in_flight_count_ < MAX_PARALLELISM) || abort_; });
+    if (!abort_) {
+        if (success) {
+            in_flight_count_++;
+        } else {
+            abort_ = true;
+        }
+    }
+    lock.unlock();
+    return !abort_;
+}
+
+void
+unstaging_state::notify_unstage_complete()
+{
+    std::lock_guard lock(mutex_);
+    in_flight_count_--;
+    cv_.notify_one();
+}
+
+void
+unstaging_state::notify_unstage_error()
+{
+    std::lock_guard lock(mutex_);
+    abort_ = true;
```

Comment:
I don't think is necessary at the moment. If `abort_` is set, the  `(in_flight_count_ < MAX_PARALLELISM) || abort_` condition will always evaluate to true and then the unstaging of any further mutations gets cancelled, so the value of `in_flight_count_` doesn't matter. But I will reduce it just in case there are any changes in the future and it gets missed